### PR TITLE
RFC: Fix converting types for UNION queries (may produce LOGICAL_ERROR)

### PR DESCRIPTION
**Pull request by azat**
Changelog category (leave one):
- Bug Fix (user-visible misbehaviour in official stable or prestable release)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix converting types for UNION queries (may produce LOGICAL_ERROR)

CI founds [1]:

    2022.02.20 15:14:23.969247 [ 492 ] {} <Fatal> BaseDaemon: (version 22.3.1.1, build id: 6082C357CFA6FF99) (from thread 472) (query_id: a5187ff9-962a-4e7c-86f6-8d48850a47d6)
    (query: SELECT 0., round(avgWeighted(x, y)) FROM (SELECT toDate(toDate('214748364.8', '-922337203.6854775808', '-0.1', NULL) - NULL, 10.000100135803223, '-2147483647'), 255 AS x, -2147483647 AS y UNION ALL SELECT y, NULL AS x, 2147483646 AS y)) Received signal Aborted (6)

  [1]: https://s3.amazonaws.com/clickhouse-test-reports/0/26d0e5438c86e52a145aaaf4cb523c399989a878/fuzzer_astfuzzerdebug,actions//report.html

The problem is that subqueries returns different headers:
- first query  -- x, y
- second query -- y, x

Fix this, by matching header by names.

### CODE REVIEW

**Comment by azat** [comment]
@novikd can you please take a look?

File: src/Interpreters/SelectQueryOptions.h [hunk 1]
```diff
@@ -31,6 +31,8 @@ struct SelectQueryOptions
     bool only_analyze = false;
     bool modify_inplace = false;
     bool remove_duplicates = false;
+    /// This is required for UNION to match headers correctly.
+    bool strict_order = false;
```
Comment (novikd) [inline src/Interpreters/SelectQueryOptions.h:35]: Variable name is unclear. I'd suggest something like `reorder_columns_as_reqiured_header`.

File: src/Interpreters/SelectQueryOptions.h [hunk 2]
```diff
@@ -96,6 +98,12 @@ struct SelectQueryOptions
         return *this;
     }
 
+    SelectQueryOptions & strictOrder(bool value = true)
```
Comment (novikd) [inline src/Interpreters/SelectQueryOptions.h:101]:
```suggestion
    SelectQueryOptions & reorderColumns(bool value = true)
```

**Review by novikd — COMMENTED** [review]
It looks as a not bad workaround for a very rare case. If we merge it, we'll need to support it and as far as I know we plan to remove `TreeRewriter` after new identifier analyser is ready. So I'd prefer to move fix to another place if it's possible.

Also I see that request below already works now:
```
SELECT x, y FROM n1 UNION ALL SELECT y, x FROM n1
```
I didn't find where header is changed in this case. Maybe we can make code for this logic more general to support your case?
@azat What do you think?

**Comment by novikd** [comment]
@Mergifyio update

**Comment by azat** [comment]
>SELECT x, y FROM n1 UNION ALL SELECT y, x FROM n1

This is the problem query `SELECT avgWeighted(x, y) FROM (SELECT NULL, 255 AS x, 1 AS y UNION ALL SELECT y, NULL AS x, 1 AS y)`, and AFAICS it does not work and leads to SIGSEGV

>It looks as a not bad workaround for a very rare case. If we merge it, we'll need to support it and as far as I know we plan to remove TreeRewriter after new identifier analyser is ready.

Once it will be removed the problem may (or may not) go away.

>So I'd prefer to move fix to another place if it's possible.

Any suggestions? The problem is that it reorders columns while UNION matches by position.

**Comment by azat** [comment]
>[Stress test (debug, actions) — Backward compatibility check](https://s3.amazonaws.com/clickhouse-test-reports/34775/e3c5d110c147d35b809c657eddf657e1362fc486/stress_test__debug__actions_.html)

- #35503

*[Build page](https://s3.amazonaws.com/clickhouse-builds/34775/e3c5d110c147d35b809c657eddf657e1362fc486/clickhouse_build_check_(actions)/report.html)*

**Comment by novikd** [comment]
Test failures seem to be unrelated to this PR. I created an issue #37569

**Comment by novikd** [comment]
I had to revert this PR, because stress test failures were definitely introduced by this PR. @azat please take a look at #37569

**Comment by azat** [comment]
There are more regressions with this patch that it solves, I'm working on a new one right now.

**Pull request merged**